log4j-to-jul/src/main/java/org/apache/logging/log4j/tojul/LazyLog4jLogRecord.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache license, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the license for the specific language governing permissions and
+ * limitations under the license.
+ */
+package org.apache.logging.log4j.tojul;
+
+import java.util.logging.Level;
+import java.util.logging.LogRecord;
+import org.apache.logging.log4j.util.StackLocatorUtil;
+
+/**
+ * Extension of {@link java.util.logging.LogRecord} with lazy get source related methods based on Log4j's {@link StackLocatorUtil#calcLocation(String)}.
+ */
+final class LazyLog4jLogRecord extends LogRecord {
+
+    private static final long serialVersionUID = 6798134264543826471L;
+
+    // parent class LogRecord already has a needToInferCaller but it's private
+    private transient boolean inferCaller = true;
+
+    private final String fqcn;
+
+    LazyLog4jLogRecord(String fqcn, Level level, String msg) {
+        super(level, msg);
+        this.fqcn = fqcn;
+    }
+
+    @Override
+    public String getSourceClassName() {
+        if (inferCaller) {
+            inferCaller();
+        }
+        return super.getSourceClassName();
+    }
+
+    @Override
+    public String getSourceMethodName() {
+        if (inferCaller) {
+            inferCaller();
+        }
+        return super.getSourceMethodName();
+    }
+
+    private synchronized void inferCaller() {
+        StackTraceElement location = StackLocatorUtil.calcLocation(fqcn);
+        setSourceClassName(location.getClassName());
+        setSourceMethodName(location.getMethodName());
```

Comment:
I think we may need some null checks here in case `fqcn` is null, or the resulting `location` is null. In either case we should probably explicitly set the `sourceClassName` and `sourceMethodName` to `null`.

Reply:
Done, hope looks about right like that?

Reply:
:+1: